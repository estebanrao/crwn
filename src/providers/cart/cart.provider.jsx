import React, { createContext, useState, useEffect } from 'react';

import {
  addItemToCart,
  removeItemFromCart,
  filterItemsFromCart,
  getCartItemsCount,
  getCartTotal,
} from './cart.utils';

export const CartContext = createContext({
  hidden: true,
  toggleHidden: () => {},
  cartItems: [],
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
  cartItemsCount: 0,
  cartItemsTotal: 0,
});

const CartProvider = ({ children }) => {
  const [hidden, setHidden] = useState(true);
  const [cartItems, setCartItems] = useState([]);
  const [cartItemsCount, setCartItemsCount] = useState(0);
  const [cartItemsTotal, setCartItemsTotal] = useState(0);

  const addItem = (item) => setCartItems(addItemToCart(cartItems, item));
  const removeItem = (item) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const toggleHidden = () => setHidden(!hidden);
  const clearItemFromCart = (item) =>
    setCartItems(filterItemsFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setCartItemsTotal(getCartTotal(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        hidden,
        toggleHidden,
        cartItems,
        addItem,
        removeItem,
        cartItemsCount,
        clearItemFromCart,
        cartItemsTotal,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
