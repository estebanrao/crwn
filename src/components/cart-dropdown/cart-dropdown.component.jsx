import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Link } from 'react-router-dom';

import CustomButton from '../custom-button/custom-button.component';
import CartItem from '../cart-item/cart-item.component';
import { selectCartItems } from '../../redux/cart/cart.selectors';
import { toggleCartHidden } from '../../redux/cart/cart.actions';

import {
  CartDropDownContainer,
  CartItemsContainer,
  EmptyMessage,
} from './cart-dropdown.styles';

export const CartDropdown = ({ cartItems, dispatch }) => {
  return (
    <CartDropDownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessage>Your cart is empty</EmptyMessage>
        )}
      </CartItemsContainer>
      <CustomButton
        as={Link}
        to="/checkout"
        onClick={() => dispatch(toggleCartHidden())}
      >
        GO TO CHECKOUT
      </CustomButton>
    </CartDropDownContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

export default connect(mapStateToProps)(CartDropdown);
