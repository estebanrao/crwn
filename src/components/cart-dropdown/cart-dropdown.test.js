import React from 'react';
import { shallow } from 'enzyme';

import { CartDropdown } from './cart-dropdown.component';
import CartItem from '../cart-item/cart-item.component';
import CustomButton from '../custom-button/custom-button.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions';

const mockProps = {
  cartItems: [],
  dispatch: jest.fn(),
};
let wrapper;

beforeEach(() => {
  wrapper = shallow(<CartDropdown {...mockProps} />);
});

describe('CartDropdown', () => {
  it('should render CardDropdown component correctly with empty message', () => {
    expect(wrapper.debug()).toMatchSnapshot(); // First snapshot created with EmptyMessage element
    expect(wrapper.find('EmptyMessage').length).toEqual(1);
  });

  it('should render CardDropdown component correctly with cart items passed in props', () => {
    wrapper.setProps({ cartItems: [{ id: 'a' }, { id: 'b' }] });
    expect(wrapper.debug()).toMatchSnapshot(); // Second snapshot created with CartItem elements
    expect(wrapper.find(CartItem).length).toEqual(2);
  });

  it('should toggle cart hidden on checkout button click', () => {
    wrapper.find(CustomButton).simulate('click');
    expect(mockProps.dispatch).toHaveBeenCalledWith(toggleCartHidden());
  });
});
