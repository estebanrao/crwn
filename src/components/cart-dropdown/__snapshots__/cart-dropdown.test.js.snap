// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown should render CardDropdown component correctly with cart items passed in props 1`] = `
"<CartDropDownContainer>
  <CartItemsContainer>
    <Memo(CartItem) item={{...}} />
    <Memo(CartItem) item={{...}} />
  </CartItemsContainer>
  <CustomButton as={{...}} to=\\"/checkout\\" onClick={[Function: onClick]}>
    GO TO CHECKOUT
  </CustomButton>
</CartDropDownContainer>"
`;

exports[`CartDropdown should render CardDropdown component correctly with empty message 1`] = `
"<CartDropDownContainer>
  <CartItemsContainer>
    <EmptyMessage>
      Your cart is empty
    </EmptyMessage>
  </CartItemsContainer>
  <CustomButton as={{...}} to=\\"/checkout\\" onClick={[Function: onClick]}>
    GO TO CHECKOUT
  </CustomButton>
</CartDropDownContainer>"
`;
